pkg/pillar/hypervisor/containerd.go

```diff
@@ -338,3 +338,28 @@ func (ctx ctrdContext) VirtualTPMTerminate(domainName string) error {
 func (ctx ctrdContext) VirtualTPMTeardown(domainName string) error {
 	return fmt.Errorf("not implemented")
 }
+
+func (ctx ctrdContext) createXenToolsImage() error {
+	ctrdUserCtx, done, err := ctx.ctrdClient.CtrNewUserServicesCtxWithLease()
+	if err != nil {
+		return fmt.Errorf("couldn't create user context with lease: %v.", err)
+	}
+	defer done()
+
+	// check if the image xen-tools is available
+	imageExists, err := ctx.ctrdClient.CtrImageExists(ctrdUserCtx, "xen-tools:local")
+	if err != nil {
+		return fmt.Errorf("couldn't check if image xen-tools exists: %v.", err)
+	}
+	if imageExists {
```

Comment:
Does this image include the shim-VM linux kernel in addition to the xen-tools user-space?

Then there is an interesting question about EVE updates (and also about how this works today).
If a device is running EVE version X, and you deploy container A, it gets the Linux kernel from version X.
Then you update EVE to version Y. Do you expect container A to restart with the new Linux kernel from EVE version Y or not? (and I don't know what happens today for sure - we already have the shim-VM rootfs in /persist/vault/volumes but what about the kernel?)
Then you deploy container B. Do you expect it to get the version X or Y kernel?

In addition to the kernel, I assume we want to be able to deploy fixes (e.g., to the dhcp and mount scripts) in the shim VM and have those be used by existing and future deployed containers.

Reply:
Doesn't the shim-VM use the EVE kernel? I had thought it just mounts in the kernel from EVE itself?

Reply:
I don't know about the kernel.

But user-space bug fixes matter as well.

Reply:
> Doesn't the shim-VM use the EVE kernel?

The kernel that runs the container is exactly the same as the host. A container can just have some file overlays + limits in access to some resources. Runtime is shared with the host in any case. So no risk of Kernel version mismatch. So the only similar possible problem I could imagine would be the Linux modules that we can "repopulate" into image as we do mount them here: https://github.com/lf-edge/eve/blob/019f18977cea01b321f512e2064d7c4162114328/pkg/pillar/containerd/oci.go#L186 But I'm not sure it's really the case.

Reply:
Quite true.

To some degree, I don't think we have asked these questions in a while. In the immediate sense, @europaul 's PR does not change how anything works now. Same filesystem, same kernel, same everything. It just sets up the filesystem path correctly, rather than sharing (and trouncing) like it does now. And relies on containerd to do it correctly. So I think we should not hold this up for that.

We probably should have a proper design discussion around the issues you have raised. What are the behaviours we are trying to achieve? What are potential challenges? How should we restructure it?

Reply:
Doesn't it change the behavior of container B which is deployed after EVE is updated to version Y?
Or does that currently reuse the user-space image that was created when container A was deployed earlier? (on EVE version X.)

Reply:
If I'm correct, we recreate the "base" image on each device reboot now (which I find odd, by the way, as we could do it in build time). But I'm not 100% sure I understand the code correctly...

Reply:
> Doesn't it change the behavior of container B

Not really. All of this is using xen-tools. Whether we use the mounted one from the root filesystem, or whether we copy it to an image and then mount that, we still are using whatever is on the rootfs. 

* Current: we are using xen-tools:1, it mounts the directory from running xen-tools, launches container B. We upgrade EVE, including xen-tools, which includes xen-tools:2. It mounts the directory from running xen-tools (clueless as to what version it is) and launches B. B just got a newer version of xen-tools.
* Proposed: we are using xen-tools:1, it creates an image from xen-tools, launches container B. We upgrade EVE, including xen-tools, which includes xen-tools:2. It creates an image from xen-tools (clueless as to what version it is), and launches B. B just got a newer version of xen-tools.

The case I can see (maybe this is what you were referring to, Erik?) is if, after the upgrade, it looks in containerd and says, "oh, wait I already have xen-tools, so I won't bother creating it from the filesystem." Then container B is stuck on xen-tools:1, because it was saved in containerd.

Is that what you were asking about? I am not sure if it works this way or the way I described in "Proposed" above. @europaul can you explain?

Reply:
> If I'm correct, we recreate the "base" image on each device reboot now

Not sure, Paul would know. If so, that would resolve the issue Erik raised in the previous comment (if I understood him).

>  (which I find odd, by the way, as we could do it in build time).

That would be even nicer. If we can find a way...

Reply:
> > Doesn't it change the behavior of container B
> 
> Not really. All of this is using xen-tools. Whether we use the mounted one from the root filesystem, or whether we copy it to an image and then mount that, we still are using whatever is on the rootfs.
> 
> * Current: we are using xen-tools:1, it mounts the directory from running xen-tools, launches container B. We upgrade EVE, including xen-tools, which includes xen-tools:2. It mounts the directory from running xen-tools (clueless as to what version it is) and launches B. B just got a newer version of xen-tools.
> * Proposed: we are using xen-tools:1, it creates an image from xen-tools, launches container B. We upgrade EVE, including xen-tools, which includes xen-tools:2. It creates an image from xen-tools (clueless as to what version it is), and launches B. B just got a newer version of xen-tools.
> 
> The case I can see (maybe this is what you were referring to, Erik?) is if, after the upgrade, it looks in containerd and says, "oh, wait I already have xen-tools, so I won't bother creating it from the filesystem." Then container B is stuck on xen-tools:1, because it was saved in containerd.
> 
> Is that what you were asking about? I am not sure if it works this way or the way I described in "Proposed" above. @europaul can you explain?

@deitch, yes that is what I'm asking about. AFAICT the new code checks if the image exists in containerd and if not it will never create a new version of that image. Sounds like this is a change.
FWIW we could name that xen-tools image based on the eve-version string to make this behave the same as today (if we understand the current behavior per above.) But that would require having code to delete the unreferenced old ones from containerd.

Reply:
Well, not bad. I had to walk through the whole path to get it, but I did in the end. 😄 

@europaul can you address his issue? It is a valid one. We want to resolve this issue you have raised, but without changing behaviour.

Reply:
First of all thanks you for all the comments, folks, I really appreciate it! I'm gonna start with this conversation since it has the most fundamental questions.

`xen-tools:local` image is supposed to be created on the pillar startup, unless it already exists in the containerd image store. So it's not recreated after a simple reboot.
As far as my understanding goes, when EVE is upgraded the partition B is clean - like on the first ever start of EVE - so there will be no images in the containerd image store. That means that `xen-tools:local` image will be recreated from the rootfs of the new version of xen-tools.

It would be great if we could put a xen-tools image into the containerd's image store at build time, but afaiu @deitch said that would require changes on the linuxkit front (which I'm happy to make if needs be).